### ISSUE TITLE
Aws::S3::Errors::InvalidArgument with SSE-C, Aws::S3::Object#upload_file, and non-multipart-upload

### ISSUE THREAD

**Issue by ttsumibishi** [aws/aws-sdk-ruby#1302]
I'm implementing the option to use SSE-C with a piece of software, using a client-provided AES-256 key. That part all works fine. However, when the file I'm uploading using AWS::S3::Object#upload_file is lower than the multipart threshold value (default 15M), I receive the error as shown below.

`Aws::S3::Errors::InvalidArgument Server Side Encryption with Customer provided key is incompatible with the encryption method specified
`

This error is thrown when I am using SSE-C with a 55k file, 10M file, 14.9M file, or 15728639 byte file. I created the files using 'dd if=/dev/urandom of=test_file bs=1 count=X' (X being how many bytes I wanted - did this for granularity as I approached threshold, but before that I did by 1MB chunks). Once I hit the threshold value and the SDK decided to use multi-part uploads, everything worked fine.

To test this further, I set the threshold to '1', and every file above then uploaded fine, where it would fail before it multipart upload wasn't used.

This was against us-west-2 with Ruby 2.1 and version 2.5.1 of the SDK (what's been imported into RubyGem-aws-sdk-resources 2.x MV).

**Comment by ttsumibishi** [comment]
Tested against 2.6.3 (public) with a completely stripped down version of my code (so I can post here). Removed access/secret keys, bucket name from output.

Also, with 2.6.3 I tested with various file sizes (55k, 1M, 10M, 20M, 100M) and multipart thresholds (default, 5M, 1 byte), and could not successfully upload a file with SSE-C.

No uploads have been successful with 2.6.3.

## **Code**

```
#!/usr/bin/ruby

require 'openssl'
require 'aws-sdk'
require 'logger'
require 'base64'
require 'socket'
require 'fileutils'
require 'pp'

# Just for testing
symmetric_key = OpenSSL::Cipher::AES256.new(:CBC).random_key

client = Aws::S3::Client.new(
      region:            'us-west-2',
      access_key_id:     '[KNOPE]',
      secret_access_key: '[KNOPE]',
      log_level:         :debug,
      logger:            Logger.new($stdout)
  )

s3 = Aws::S3::Resource.new(client: client, http_wire_trace: true)

File.open(ARGV[0]) do |file|
    f = File.basename(file)
    obj = s3.bucket('[KNOPE]').object(f)
    pp obj
    obj.upload_file(file, { content_type: 'application/x-xz',
                            storage_class: 'STANDARD',
                            server_side_encryption: 'AES256',
                            sse_customer_algorithm: 'AES256',
                            sse_customer_key: symmetric_key,
                            sse_customer_key_md5: Digest::MD5.base64digest(symmetric_key),
                            multipart_threshold: 1 })
end
```

## **Error**

```
[ttsumi@freeside testing]$ ./s3.rb test_file_20m.tar.xz
#<Aws::S3::Object bucket_name="[KNOPE]", key="test_file_20m.tar.xz">
D, [2016-09-24T13:34:40.074772 #23771] DEBUG -- : [Aws::S3::Client 400 0.102537 0 retries] create_multipart_upload(bucket:"[KNOPE]",content_type:"application/x-xz",key:"test_file_20m.tar.xz",server_side_encryption:"AES256",storage_class:"STANDARD",sse_customer_algorithm:"AES256",sse_customer_key:"[FILTERED]",sse_customer_key_md5:"[KNOPE]") Aws::S3::Errors::InvalidArgument Server Side Encryption with Customer provided key is incompatible with the encryption method specified

/home/ttsumi/.gem/ruby/gems/aws-sdk-core-2.6.3/lib/seahorse/client/plugins/raise_response_errors.rb:15:in `call': Server Side Encryption with Customer provided key is incompatible with the encryption method specified (Aws::S3::Errors::InvalidArgument)
        from /home/ttsumi/.gem/ruby/gems/aws-sdk-core-2.6.3/lib/aws-sdk-core/plugins/s3_sse_cpk.rb:19:in `call'
        from /home/ttsumi/.gem/ruby/gems/aws-sdk-core-2.6.3/lib/aws-sdk-core/plugins/s3_dualstack.rb:23:in `call'
        from /home/ttsumi/.gem/ruby/gems/aws-sdk-core-2.6.3/lib/aws-sdk-core/plugins/s3_accelerate.rb:33:in `call'
        from /home/ttsumi/.gem/ruby/gems/aws-sdk-core-2.6.3/lib/aws-sdk-core/plugins/param_converter.rb:20:in `call'
        from /home/ttsumi/.gem/ruby/gems/aws-sdk-core-2.6.3/lib/seahorse/client/plugins/response_target.rb:21:in `call'
        from /home/ttsumi/.gem/ruby/gems/aws-sdk-core-2.6.3/lib/seahorse/client/request.rb:70:in `send_request'
        from /home/ttsumi/.gem/ruby/gems/aws-sdk-core-2.6.3/lib/seahorse/client/base.rb:207:in `block (2 levels) in define_operation_methods'
        from /home/ttsumi/.gem/ruby/gems/aws-sdk-resources-2.6.3/lib/aws-sdk-resources/services/s3/multipart_file_uploader.rb:52:in `initiate_upload'
        from /home/ttsumi/.gem/ruby/gems/aws-sdk-resources-2.6.3/lib/aws-sdk-resources/services/s3/multipart_file_uploader.rb:43:in `upload'
        from /home/ttsumi/.gem/ruby/gems/aws-sdk-resources-2.6.3/lib/aws-sdk-resources/services/s3/file_uploader.rb:32:in `upload'
        from /home/ttsumi/.gem/ruby/gems/aws-sdk-resources-2.6.3/lib/aws-sdk-resources/services/s3/object.rb:251:in `upload_file'
        from ./test.rb:37:in `block in <main>'
        from ./test.rb:33:in `open'
        from ./test.rb:33:in `<main>'
```

**Comment by ttsumibishi** [comment]
Got it. Digging in a bit (read: lot) more, it was unclear in the public AWS documentation that if I supply the sse_customer_algorithm attribute, it is mutually exclusive (and will throw an error) with the server_side_encryption attribute. The docs only state that if I want to use AMS-managed keys, to set "server_side_encryption" attribute to something (like AES256 or KMS), and if I want to supply customer keys, to add the three sse_customer_* attributes. It was my assumption that the server_side_encryption option was just telling the upload that I want to use server side encryption vs client side, not that it was solely to be used for AWS-managed keys.

"Server side encryption with customer provided key is incompatible with the encryption method specified." is very confusing, as that's what SSE-C is defined as, and since it only supports AES256 (which I thought to be the encryption method), it was very misleading.

Taking out the server_side_encryption attribute fixes it.

What's kind of funny is with 2.5.1, if I set multipart threshold to 1, it worked regardless of this parameter being in the options hash.

Closing this for now. I'll find the right channel to fix the public documentation and maybe clarify the message.
